### ISSUE TITLE
Segfault running shape inference with `data_prop=True` on model with nested function.

### ISSUE THREAD

**Issue by BowenBao** [onnx/onnx#5212]
# Bug Report

### Is the issue related to model conversion?

No, checker passed. inliner passed. 
Same shape inference call passed on inlined model.

### Describe the bug

As title

### System information

ONNX version: 1.15 dbeee2465b8628ddf742e8ff52144b424f3c45db

### Reproduction instructions
Model can be downloaded [here](https://microsoft-my.sharepoint.com/:u:/p/bowbao/EZVQDNP6-wpBqmls_ScqhgYB9iQJe_P0vbsZct55Gtw7MQ?e=9MBbVO).
```
import onnx
import onnx.shape_inference
import onnx.inliner

model = onnx.load("maf_mnist.onnx")

onnx.checker.check_model(model)
print("Pass checker!")

# Works fine
onnx.shape_inference.infer_shapes(model, data_prop=False)
print("Success with data_prop=False !")

inlined_model = onnx.inliner.inline_local_functions(model)
print("Success with inlining!")
onnx.shape_inference.infer_shapes(inlined_model, data_prop=True)
print("Success with data_prop=True on inlined model !")

# Segfaults
onnx.shape_inference.infer_shapes(model, data_prop=True)

```

**Comment by jcwchen** [comment]
Hi @BowenBao,
Thanks for report. After several experiments, segfault starts to happen after this PR: https://github.com/onnx/onnx/pull/5066. Will investigate more.

**Comment by BowenBao** [comment]
Got another repro for [gpt2 tiny](https://microsoft-my.sharepoint.com/:u:/p/bowbao/EZMzLywywzNKikgZWwJ7I24Bp5L28aDU9WHwdCKXiJ9j-Q?e=zzhbji)

This one segfaults with just shape inference
```
import onnx
import onnx.shape_inference
import onnx.inliner
import onnxscript

model = onnx.load("maf_gpt2_tiny.onnx")
onnx.shape_inference.infer_shapes(model)
print("Pass shape inference!")
```

It doesn't pass checker though, but I think it's because we weren't able to use shape_inference to fill in some missing output shapes at `onnxscript/function_libs/torch_lib/graph_building.py`.

**Comment by gramalingam** [comment]
May be worth checking the model-generation part. I tried the gpt2 tiny example above. I see it has functions like "aten_rsub" which uses an attribute-parameter alpha, but the function's attribute list is empty.

**Comment by gramalingam** [comment]
> May be worth checking the model-generation part. I tried the gpt2 tiny example above. I see it has functions like "aten_rsub" which uses an attribute-parameter alpha, but the function's attribute list is empty.

If it is generated by onnxscript, or exporter, need to find why the attribute-parameter list for the function is empty, even though it has one.

**Comment by gramalingam** [comment]
On further thought, I think it must be an attribute parameter with a default value. I think the issue is that the checker and shape-inference have not been updated to handle attribute-parameters with default-values, which I pointed out in the original PR. It was supposed to be done in a subsequent PR.

**Comment by BowenBao** [comment]
Reopen to keep tracking root cause ..

Hmm, applying https://github.com/onnx/onnx/pull/5219#issuecomment-1544376547 still segfaults on my end, but maybe I updated a few things in the model that is different from one that was shared. Resharing the[ latest one](https://microsoft-my.sharepoint.com/:u:/p/bowbao/Ebtt_SQsuaFDotGl5p0sIZcBAQzdDCZ4BBBI5MEFCQOm3A?e=l3Trvy) . 

I have a few observations so far and some guesses.

It looks like shape inference for `aten_view_1` in `transformers.pytorch_utils.Conv1D(L__self___h_0_attn_c_attn)` stopped in the middle. An error was captured and added to the `inference_errors` list, and shape inference continues to the next function node.
```
# printf logs
Caught InferenceError for node aten_view_1 error msg [ShapeInferenceError] (op_type:Cast, node name: n0): [ShapeInferenceError] Data for input  0 already exists.
Adding inference error for node aten_view_1
```

https://github.com/onnx/onnx/blob/065c103533c6cef70de2f23152fce67a69806600/onnx/shape_inference/implementation.cc#L413-L423

Due to this, the next op received nullptr for its input type. It propagated until reaching a `Shape` node which segfaults.

I wonder if this is the cascading result of the attribute issue or something else?

**Comment by gramalingam** [comment]
Weird, I didn't close this issue, it looks like it got closed automatically when the other PR was merged. Anyway, your explanation is what I saw in the earlier model I looked at (an attribute with default-value was not handled in an aten function, producing an inference error, which caused nullptr for input types in subsequent nodes). I will take a look at the new model you attached, thanks!

**Comment by gramalingam** [comment]
Ok, I was trying shape-inference without data-propagation (since that was failing in your original simplified example). Now, I tried data-propagation and that did fail. This looks like an independent problem. I added this partial fix [here](https://github.com/gramalingam/onnx/pull/4/files) ... and, with that, it seems to work for me even when I used data-propagation.

The issue I ran into was that the same map was being used for data-propagation across multiple function-calls, causing a name-collision. We need to use different maps when analysing a function call. The fix is only partial: it avoids incorrectness. But for more precise information, we need to copy relevant information from calling scope to called scope and back (for the inputs and outputs). I will try that later.
